### PR TITLE
Feature/google-webservices-direction

### PR DESCRIPTION
**Pull request by Mehrdadmaskull**
Added direction data model and an API call to use Google Directions

### CODE REVIEW

File: lib/service/map_direction.dart [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    if (origin.isEmpty || destination.isEmpty) {
+      throw("Origin or destination is empty");
+    }
+    if (_sessionToken == null) {
+      _sessionToken = uuid.v4();
+    }
+    
+    String baseURL = 'https://maps.googleapis.com/maps/api/directions/json';
+
+    String request = '$baseURL?origin=$origin&destination=$destination&key=$PLACES_API_KEY&sessiontoken=$_sessionToken';
```
Comment (MrJCipherButtles) [inline lib/service/map_direction.dart]: PLACES_API_KEY should be replaced with DIRECTIONS_API_KEY

File: lib/service/map_direction.dart [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import 'package:concordi_around/model/list_item.dart';
+import 'package:dio/dio.dart';
+import 'package:uuid/uuid.dart';
+
+import '../credential.dart';
+
+class MapDirection {
+
+  var uuid = Uuid();
+  String _sessionToken;
```
Comment (MrJCipherButtles) [inline lib/service/map_direction.dart]: The concept of sessionTokens do not apply in google web service direction API requests. This should be removed.

File: lib/service/map_direction.dart (hunk 1 above)
Comment (MrJCipherButtles) [reply lib/service/map_direction.dart]: Also remove sessiontoken parameter here as per above

File: lib/service/map_direction.dart [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+    // Google Directions API requests
+  Future<List<ListItem>> _getDirection(String origin, String destination) async {
+    if (origin.isEmpty || destination.isEmpty) {
+      throw("Origin or destination is empty");
+    }
+    if (_sessionToken == null) {
+      _sessionToken = uuid.v4();
+    }
```
Comment (MrJCipherButtles) [inline lib/service/map_direction.dart]: As per above, sessions should not be used for direction API calls

File: lib/service/map_direction.dart [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+import '../credential.dart';
+
+class MapDirection {
+
+  var uuid = Uuid();
+  String _sessionToken;
+
+    // Google Directions API requests
+  Future<List<ListItem>> _getDirection(String origin, String destination) async {
```
Comment (MrJCipherButtles) [inline lib/service/map_direction.dart]: You are missing a travel mode parameter

File: lib/service/map_direction.dart (hunk 1 above)
Comment (MrJCipherButtles) [reply lib/service/map_direction.dart]: Add the [mode](https://developers.google.com/maps/documentation/directions/intro#TravelModes) parameter

File: lib/model/direction.dart [hunk 5]
```diff
@@ -0,0 +1,216 @@
+class Direction {
+  List<Routes> routes;
+
+  Direction({this.routes});
+
+  Direction.fromJson(Map<String, dynamic> json) {
+    if (json['routes'] != null) {
+      routes = new List<Routes>();
```
Comment (MrJCipherButtles) [inline lib/model/direction.dart]: As per Effective Dart, [DON’T use new.](https://dart.dev/guides/language/effective-dart/usage#dont-use-new)

File: lib/model/direction.dart [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+      json['routes'].forEach((v) {
+        routes.add(new Routes.fromJson(v));
```
Comment (MrJCipherButtles) [inline lib/model/direction.dart]: As per Effective Dart, [DON’T use new.](https://dart.dev/guides/language/effective-dart/usage#dont-use-new)

File: lib/model/direction.dart [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+      });
+    }
+  }
+
+  Map<String, dynamic> toJson() {
+    final Map<String, dynamic> data = new Map<String, dynamic>();
```
Comment (MrJCipherButtles) [inline lib/model/direction.dart]: As per Effective Dart, [DON’T use new.](https://dart.dev/guides/language/effective-dart/usage#dont-use-new)

File: lib/model/direction.dart [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+      this.distance,
+      this.startLocation,
+      this.endLocation,
+      this.startAddress,
+      this.endAddress});
+
+  Routes.fromJson(Map<String, dynamic> json) {
+    summary = json['summary'];
+    if (json['legs'] != null) {
+      legs = new List<Legs>();
```
Comment (MrJCipherButtles) [inline lib/model/direction.dart]: As per Effective Dart, [DON’T use new.](https://dart.dev/guides/language/effective-dart/usage#dont-use-new)

File: lib/model/direction.dart [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+      json['legs'].forEach((v) {
+        legs.add(new Legs.fromJson(v));
```
Comment (MrJCipherButtles) [inline lib/model/direction.dart]: As per Effective Dart, [DON’T use new.](https://dart.dev/guides/language/effective-dart/usage#dont-use-new)

File: lib/model/direction.dart [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+      });
+    }
+    duration = json['duration'] != null
+        ? new Duration.fromJson(json['duration'])
```
Comment (MrJCipherButtles) [inline lib/model/direction.dart]: As per Effective Dart, [DON’T use new.](https://dart.dev/guides/language/effective-dart/usage#dont-use-new)

File: lib/model/direction.dart [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+        : null;
+    distance = json['distance'] != null
+        ? new Duration.fromJson(json['distance'])
```
Comment (MrJCipherButtles) [inline lib/model/direction.dart]: As per Effective Dart, [DON’T use new.](https://dart.dev/guides/language/effective-dart/usage#dont-use-new)

File: lib/model/direction.dart [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+        : null;
+    startLocation = json['start_location'] != null
+        ? new LocationModel.fromJson(json['start_location'])
```
Comment (MrJCipherButtles) [inline lib/model/direction.dart]: As per Effective Dart, [DON’T use new.](https://dart.dev/guides/language/effective-dart/usage#dont-use-new)

File: lib/model/direction.dart [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+        : null;
+    endLocation = json['end_location'] != null
+        ? new LocationModel.fromJson(json['end_location'])
```
Comment (MrJCipherButtles) [inline lib/model/direction.dart]: As per Effective Dart, [DON’T use new.](https://dart.dev/guides/language/effective-dart/usage#dont-use-new)

File: lib/model/direction.dart [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+        : null;
+    startAddress = json['start_address'];
+    endAddress = json['end_address'];
+  }
+
+  Map<String, dynamic> toJson() {
+    final Map<String, dynamic> data = new Map<String, dynamic>();
```
Comment (MrJCipherButtles) [inline lib/model/direction.dart]: As per Effective Dart, [DON’T use new.](https://dart.dev/guides/language/effective-dart/usage#dont-use-new)

File: lib/model/direction.dart [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+}
+
+class Legs {
+  List<Steps> steps;
+
+  Legs({this.steps});
+
+  Legs.fromJson(Map<String, dynamic> json) {
+    if (json['steps'] != null) {
+      steps = new List<Steps>();
```
Comment (MrJCipherButtles) [inline lib/model/direction.dart]: As per Effective Dart, [DON’T use new.](https://dart.dev/guides/language/effective-dart/usage#dont-use-new)

File: lib/model/direction.dart [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+      json['steps'].forEach((v) {
+        steps.add(new Steps.fromJson(v));
```
Comment (MrJCipherButtles) [inline lib/model/direction.dart]: As per Effective Dart, [DON’T use new.](https://dart.dev/guides/language/effective-dart/usage#dont-use-new)

File: lib/model/direction.dart [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+      });
+    }
+  }
+
+  Map<String, dynamic> toJson() {
+    final Map<String, dynamic> data = new Map<String, dynamic>();
```
Comment (MrJCipherButtles) [inline lib/model/direction.dart]: As per Effective Dart, [DON’T use new.](https://dart.dev/guides/language/effective-dart/usage#dont-use-new)

File: lib/model/direction.dart [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+      {this.travelMode,
+      this.startLocation,
+      this.endLocation,
+      this.polyline,
+      this.duration});
+
+  Steps.fromJson(Map<String, dynamic> json) {
+    travelMode = json['travel_mode'];
+    startLocation = json['start_location'] != null
+        ? new LocationModel.fromJson(json['start_location'])
```
Comment (MrJCipherButtles) [inline lib/model/direction.dart]: As per Effective Dart, [DON’T use new.](https://dart.dev/guides/language/effective-dart/usage#dont-use-new)

File: lib/model/direction.dart [hunk 19]
```diff
@@ -0,0 +1,216 @@
...
+        : null;
+    endLocation = json['end_location'] != null
+        ? new LocationModel.fromJson(json['end_location'])
```
Comment (MrJCipherButtles) [inline lib/model/direction.dart]: As per Effective Dart, [DON’T use new.](https://dart.dev/guides/language/effective-dart/usage#dont-use-new)

File: lib/model/direction.dart [hunk 20]
```diff
@@ -0,0 +1,216 @@
...
+        : null;
+    polyline = json['polyline'] != null
+        ? new PolylineModel.fromJson(json['polyline'])
```
Comment (MrJCipherButtles) [inline lib/model/direction.dart]: As per Effective Dart, [DON’T use new.](https://dart.dev/guides/language/effective-dart/usage#dont-use-new)

File: lib/model/direction.dart [hunk 21]
```diff
@@ -0,0 +1,216 @@
...
+        : null;
+    duration = json['duration'] != null
+        ? new Duration.fromJson(json['duration'])
```
Comment (MrJCipherButtles) [inline lib/model/direction.dart]: As per Effective Dart, [DON’T use new.](https://dart.dev/guides/language/effective-dart/usage#dont-use-new)

File: lib/model/direction.dart [hunk 22]
```diff
@@ -0,0 +1,216 @@
...
+
+  LocationModel({this.lat, this.lng});
+
+  LocationModel.fromJson(Map<String, dynamic> json) {
+    lat = json['lat'];
+    lng = json['lng'];
+  }
+
+  Map<String, dynamic> toJson() {
+    final Map<String, dynamic> data = new Map<String, dynamic>();
```
Comment (MrJCipherButtles) [inline lib/model/direction.dart]: As per Effective Dart, [DON’T use new.](https://dart.dev/guides/language/effective-dart/usage#dont-use-new)

File: lib/model/direction.dart [hunk 23]
```diff
@@ -0,0 +1,216 @@
...
+  String points;
+
+  PolylineModel({this.points});
+
+  PolylineModel.fromJson(Map<String, dynamic> json) {
+    points = json['points'];
+  }
+
+  Map<String, dynamic> toJson() {
+    final Map<String, dynamic> data = new Map<String, dynamic>();
```
Comment (MrJCipherButtles) [inline lib/model/direction.dart]: As per Effective Dart, [DON’T use new.](https://dart.dev/guides/language/effective-dart/usage#dont-use-new)

File: lib/model/direction.dart [hunk 24]
```diff
@@ -0,0 +1,216 @@
...
+
+  Duration({this.value, this.text});
+
+  Duration.fromJson(Map<String, dynamic> json) {
+    value = json['value'];
+    text = json['text'];
+  }
+
+  Map<String, dynamic> toJson() {
+    final Map<String, dynamic> data = new Map<String, dynamic>();
```
Comment (MrJCipherButtles) [inline lib/model/direction.dart]: As per Effective Dart, [DON’T use new.](https://dart.dev/guides/language/effective-dart/usage#dont-use-new)

File: lib/service/map_direction.dart [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+    
+    String baseURL = 'https://maps.googleapis.com/maps/api/directions/json';
+
+    String request = '$baseURL?origin=$origin&destination=$destination&key=$PLACES_API_KEY&sessiontoken=$_sessionToken';
+    Response response = await Dio().get(request);
+
+    print(response.data);
+
+    final jsonResponse = json.decode(response.data);
+    Direction direction = new Direction.fromJson(jsonResponse);
```
Comment (MrJCipherButtles) [inline lib/service/map_direction.dart]: As per Effective Dart, [DON’T use new.](https://dart.dev/guides/language/effective-dart/usage#dont-use-new)

File: lib/model/direction.dart [hunk 26]
```diff
@@ -0,0 +1,216 @@
...
+    polyline = json['polyline'] != null
+        ? new PolylineModel.fromJson(json['polyline'])
+        : null;
+    duration = json['duration'] != null
+        ? new Duration.fromJson(json['duration'])
+        : null;
+  }
+
+  Map<String, dynamic> toJson() {
+    final Map<String, dynamic> data = new Map<String, dynamic>();
```
Comment (MrJCipherButtles) [inline lib/model/direction.dart]: As per Effective Dart, [DON’T use new.](https://dart.dev/guides/language/effective-dart/usage#dont-use-new)

File: lib/service/map_direction.dart (hunk 4 above)
Comment (MrJCipherButtles) [reply lib/service/map_direction.dart]: You should add mode as a ordered optional parameter with a default value of 'car' as such: `_getDirection(String origin, String destination, [String mode = 'car'])`

Also double check this functions return type as you are not returning anything. Should it be void?

Comment (MrJCipherButtles) [reply lib/service/map_direction.dart]: Also double check this functions return type as you are not returning anything. Should it be void?

File: lib/service/map_direction.dart (hunk 1 above)
Comment (Mehrdadmaskull) [reply lib/service/map_direction.dart]: Places and directions seem to be sharing one API key

Comment (Mehrdadmaskull) [reply lib/service/map_direction.dart]: Mode parameter is pending, changes from #101 is needed

**Comment by Mehrdadmaskull** [comment]
@MrJCipherButtles check the updated changes

File: lib/service/map_direction.dart (hunk 1 above)
Comment (MrJCipherButtles) [reply lib/service/map_direction.dart]: The way that the credentials.dart file was done is that. PLACES_API_KEY can differ from DIRECTIONS_API_KEY.

mode should still be part of the API request regardless of #101

**Comment by MrJCipherButtles** [comment]
> @MrJCipherButtles check the updated changes

Why did you change the credential.dart file? Keep things simple, by updating your own code with DIRECTIONS_API_KEY

**Comment by DiveshPatel** [comment]
Will driving mode be supported in this PR as per MrJCipherButtles' request?

**Comment by Mehrdadmaskull** [comment]
> Will driving mode be supported in this PR as per MrJCipherButtles' request?

Now that your PR has been approved yeah

**Comment by MrJCipherButtles** [comment]
> Will driving mode be supported in this PR as per MrJCipherButtles' request?

Google Webservices Directions supports many travel modes. We only use what we need...

**Comment by Mehrdadmaskull** [comment]
> > @MrJCipherButtles check the updated changes
> 
> Why did you change the credential.dart file? Keep things simple by updating your own code with DIRECTIONS_API_KEY. What you just did is that anywhere that was using PLACES_API_KEY had to be refactored.

There's only one API key for any API call, so why should we have two?

**Comment by MrJCipherButtles** [comment]
> > > @MrJCipherButtles check the updated changes
> > 
> > 
> > Why did you change the credential.dart file? Keep things simple by updating your own code with DIRECTIONS_API_KEY. What you just did is that anywhere that was using PLACES_API_KEY had to be refactored.
> 
> There's only one API key for any API call, so why should we have two?

Why does this have to be complicated? If you want to use the same API key for both places and directions then simply copy-paste the same key in both PLACES_API_KEY and DIRECTIONS_API_KEY. If someone wishes to separate their API keys by giving singular restrictions, then they should be able to do that. This really shouldn't be a discussion point

**Comment by MrJCipherButtles** [comment]
@Mehrdadmaskull in life if some requests a change from you, and you go out of your way to do a change that is different than what was being asked... forget it. It's simple, what you want is your PR's to be approved and merged as quickly as possible, so, it's a good policy to stick to the changes that are being asked in general.

**Comment by DiveshPatel** [comment]
Any updates on this PR since this is currently blocking 4 user stories?

**Comment by MrJCipherButtles** [comment]
@DiveshPatel I took upon myself to resolve this. Merging this into master now.

**Comment by DiveshPatel** [comment]
> @DiveshPatel I took upon myself to resolve this. Merging this into master now.

Thanks!

**Pull request merged**